src/cysignals/signals.pyx

```diff
@@ -261,3 +284,45 @@ def python_check_interrupt(sig, frame):
     ``implementation.c``.
     """
     sig_check()
+
+
+cdef void verify_exc_value():
+    """
+    Check that ``cysigs.exc_value`` is still the exception being raised
+    (and is not being handled). Clear ``cysigs.exc_value`` if not.
+    """
+    if cysigs.exc_value.ob_refcnt == 1:
+        # No other references => exception is certainly gone
+        Py_XDECREF(cysigs.exc_value)
+        cysigs.exc_value = NULL
+        return
+
+    # We consider the exception in cysigs.exc_value active, even if
+    # there is no actual exception (as returned by PyErr_Occurred).
+    # This is to support the case where the exception is temporarily
+    # disabled by a PyErr_Fetch/PyErr_Restore pair. This happens for
+    # example in Cython's __dealloc__ functions.
+
+    # There is one exception on Python 2: when an exception is
+    # referenced in sys.last_value, we know that it has been handled.
+    # We need to check this because sys.last_value "leaks" a reference
+    # to the exception.
+    if PY_MAJOR_VERSION < 3:
+        try:
+            handled = sys.last_value
+        except AttributeError:
+            pass
+        else:
+            if <PyObject*>handled is cysigs.exc_value:
+                Py_XDECREF(cysigs.exc_value)
+                cysigs.exc_value = NULL
+                return
+
+    # To be safe, we run the garbage collector because it may clear
+    # references to our exception.
```

Comment:
I'd have to think about an actual example, but couldn't an exception wind up in a reference cycle, but also be in `sys.last_value`,  in which the above case should be considered again too?  Perhaps this could install call recursively back into `verify_exc_value` with a flag set to not run garbage collection a second time.

Reply:
> I'd have to think about an actual example, but couldn't an exception wind up in a reference cycle, but also be in sys.last_value, in which the above case should be considered again too?

If the exception is `sys.last_value`, the refcount doesn't matter: we assume that the exception has been handled.

Reply:
Speaking of garbage collection, I'm not sure whether we should actually do that. If we are calling `sig_occurred` a lot of times with an active exception, that will be a massive performance loss.

It all depends how likely it is that an exception will end up in a reference cycle.

Reply:
I see what you're saying about `sys.last_value`.  At that point we don't care anymore.

Reply:
On the issue of garbage collection--it's a shame the gc module doesn't expose an API to check for a given object if it's reachable from outside the GC, without actually performing a full collection.

If nothing else the import statement should be moved outside this function--there's no reason for it to be inside.  We could also avoid an unnecessary full collection by checking `gc.get_referrers(cysigs.exc_value)`.  That doesn't necessarily guarantee that it's part of a reference cycle or that `collect()` would do anything to affect its reference count (e.g. it could be part of a cycle that includes an object with `__del__`).   But if it has no referrers tracked by the GC then obviously it's unreachable.

Reply:
Such a check wouldn't alleviate the case you mentioned though where the exception is being handled and there are other objects referring to it.  Then repeated calls to `sig_occurred` would still be slow.

Reply:
> But if it has no referrers tracked by the GC then obviously it's unreachable.

I disagree here. Not every object reference is tracked by GC. For example, Cython local variables wouldn't be and I doubt that `sys.exc_info` is.

So, I don't see how `get_referrers` would solve anything.